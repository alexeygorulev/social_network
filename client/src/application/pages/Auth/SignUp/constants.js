import keyMirrorWithPrefix from 'application/utils/keyMirrorWithPrefix';

export const FIELDS = keyMirrorWithPrefix(
  {
    LOGIN: null,
    PASSWORD: null,
    EMAIL: null,
  },
  'PARAM_TYPES_ITEM_FIELD_',
);

export const LABELS = {
  ERRORS_FIELDS: 'Обязательное поле',
  ERROR_BAD_REQUEST: 'Неправильный логин или пароль',
  ERROR_FILL_FIELDS: 'заполните, пожалуйста, все поля',
  ERROR_LOGIN: 'логин должен быть больше 3 символов и меньше 15',
  ERROR_PASSWORD: 'слишком легкий пароль',
  ERROR_PASSWORD_MAX: 'пароль должен содержать до 20 символов',
  ERROR_LOGIN: 'Пользователь с таким логином уже существует',
  ERROR_EMAIL: 'не корректный email',
  ERROR_UNIQUE_EMAIL: 'Пользователь с такой почтой уже существует',
};

export const ERROR_LABELS = {
  ERROR_LOGIN: 'Пользователь с таким логином уже существует',
  ERROR_LOGIN_MIN: 'login must be longer than or equal to 3 characters',
  ERROR_LOGIN_MAX: 'login must be shorter than or equal to 15 characters',
  ERROR_PASSWORD: 'password must be longer than or equal to 3 characters',
  ERROR_PASSWORD_MAX: 'password must be shorter than or equal to 20 characters',
  ERROR_EMAIL: 'не  корректный email',
  ERROR_UNIQUE_EMAIL: 'Пользователь с такой почтой уже существует',
};
