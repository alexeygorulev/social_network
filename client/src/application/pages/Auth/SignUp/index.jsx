import React from 'react';
import PropTypes from 'prop-types';
import { inject } from 'mobx-react';

import Component from './SignUp';

function Module(props) {
  const { store } = props;
  const { authContent } = store;
  const { signUpStore } = authContent;

  return <Component {...props} store={signUpStore} />;
}

Module.propTypes = {
  store: PropTypes.shape().isRequired,
};

export default inject('store')(Module);
