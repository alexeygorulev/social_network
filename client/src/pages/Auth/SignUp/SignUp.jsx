import s from "../Auth.module.scss";
import { Formik, Form, Field } from "formik";
import * as yup from "yup";

const validationSchema = yup.object().shape({
  email: yup
    .string()
    .email("Введите корректный email")
    .required("Обязательноe поле"),
  firstName: yup.string().required("Обязательноe поле"),
  password: yup.string().required("Обязательноe поле"),
});

const SignUp = () => {
  return (
    <div className={s.card__back}>
      <div className={s.center__wrap}>
        <div style={{ textAlign: "center" }}>
          <h4 style={{ marginBottom: 5, paddingBottom: 5 }}>Sign up</h4>
          <Formik
            initialValues={{
              email: "",
              password: "",
              firstName: "",
            }}
            validationSchema={validationSchema}
            validateOnBlur
            onSubmit={(values) => console.log(values)}
          >
            {({
              values,
              errors,
              touched,
              handleChange,
              handleBlur,
              isValid,
              handleSubmit,
              dirty,
            }) => (
              <Form>
                <div className={s.form__group}>
                  <Field
                    type="firstName"
                    name="firstName"
                    className={s.form__style}
                    placeholder="Your Name"
                    id="firstName"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.firstName}
                  />
                  {touched.firstName && errors.firstName && (
                    <p className={s.errors}>{errors.firstName}</p>
                  )}
                  <p className={s.input__icon}>
                    <span className="material-symbols-outlined">person</span>
                  </p>
                </div>
                <div className={s.form__group} style={{ marginTop: 10 }}>
                  <Field
                    type="email"
                    name="email"
                    className={s.form__style}
                    id="email"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.email}
                    placeholder="Your Email"
                  />
                  {touched.email && errors.email && (
                    <p className={s.errors}>{errors.email}</p>
                  )}
                  <p className={s.input__icon}>
                    <span className="material-symbols-outlined">
                      alternate_email
                    </span>
                  </p>
                </div>
                <div className={s.form__group} style={{ marginTop: 10 }}>
                  <Field
                    type="password"
                    name="password"
                    className={s.form__style}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.password}
                    placeholder="Your Password"
                  />
                  {touched.password && errors.password && (
                    <p className={s.errors}>{errors.password}</p>
                  )}
                  <p className={s.input__icon}>
                    <span className="material-symbols-outlined">lock</span>
                  </p>
                </div>
                <button type="submit" className={s.btn}>
                  submit
                </button>
              </Form>
            )}
          </Formik>
        </div>
      </div>
    </div>
  );
};

export default SignUp;
