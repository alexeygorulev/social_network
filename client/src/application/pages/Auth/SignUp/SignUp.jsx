import s from '../Auth.module.scss';
import { observer } from 'mobx-react';
import InputText from 'application/common/Fields/InputText';
import { useEffect } from 'react';
import { FIELDS, LABELS } from './constants';
import PropTypes from 'prop-types';


const SignUp = (props) => {
  const { store, onLogin } = props;
  const { mounted, mount, unmount, values, onBlur, onChange, addNewUser, errors, disabled, errorLabels } = store;
  useEffect(() => {
    if (!onLogin) unmount();
    return () => {};
  }, [onLogin]);
  useEffect(() => {
    if (!mounted) mount();
    return () => {
      if (mounted) unmount();
    };
  }, [mounted]);

  if (!mounted) return null;

  return (
    <div className={s.card__back}>
      <div className={s.center__wrap}>
        <div style={{ textAlign: 'center' }}>
          <h4 style={{ marginBottom: 5, paddingBottom: 5 }}>Sign up</h4>
          <div className={s.form__group}>
            <InputText
              id={FIELDS.LOGIN}
              type="text"
              onChange={onChange}
              values={values[FIELDS.LOGIN]}
              placeholder="Your Login"
              onBlur={onBlur}
            />
            {errorLabels[FIELDS.LOGIN] ? (
              <p className={s.errors}>{errorLabels[FIELDS.LOGIN]} </p>
            ) : (
              errors[FIELDS.LOGIN] !== undefined &&
              !errors[FIELDS.LOGIN] && <p className={s.errors}>{LABELS.ERRORS_FIELDS} </p>
            )}
            <p className={s.input__icon}>
              <span className="material-symbols-outlined">person</span>
            </p>
          </div>
          <div className={s.form__group} style={{ marginTop: 10 }}>
            <InputText
              id={FIELDS.EMAIL}
              type="email"
              onChange={onChange}
              values={values[FIELDS.EMAIL]}
              placeholder="Your Email"
              onBlur={onBlur}
            />
            {errorLabels[FIELDS.EMAIL] ? (
              <p className={s.errors}>{errorLabels[FIELDS.EMAIL]} </p>
            ) : (
              errors[FIELDS.EMAIL] !== undefined &&
              !errors[FIELDS.EMAIL] && <p className={s.errors}>{LABELS.ERRORS_FIELDS} </p>
            )}
            <p className={s.input__icon}>
              <span className="material-symbols-outlined">alternate_email</span>
            </p>
          </div>
          <div className={s.form__group} style={{ marginTop: 10 }}>
            <InputText
              id={FIELDS.PASSWORD}
              type="password"
              onChange={onChange}
              values={values[FIELDS.PASSWORD]}
              placeholder="Your Password"
              onBlur={onBlur}
            />
            {errorLabels[FIELDS.PASSWORD] ? (
              <p className={s.errors}>{errorLabels[FIELDS.PASSWORD]} </p>
            ) : (
              errors[FIELDS.PASSWORD] !== undefined &&
              !errors[FIELDS.PASSWORD] && <p className={s.errors}>{LABELS.ERRORS_FIELDS} </p>
            )}
            <p className={s.input__icon}>
              <span className="material-symbols-outlined">lock</span>
            </p>
          </div>
          <button disabled={disabled} onClick={addNewUser} type="submit" className={s.btn}>
            submit
          </button>
        </div>
      </div>
    </div>
  );
};

SignUp.propTypes ={
  store: PropTypes.shape().isRequired
}
export default observer(SignUp);
