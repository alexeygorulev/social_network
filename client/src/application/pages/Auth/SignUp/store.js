import { flow, getRoot, getSnapshot, types } from 'mobx-state-tree';
import { ERROR_LABELS, FIELDS, LABELS } from './constants';
import Cookies from 'react-cookie/cjs/Cookies';
import jwt from 'jwt-decode';

const cookie = new Cookies();

export const Touched = types.model(
  Object.values(FIELDS).reduce((result, item) => ({ ...result, [item]: types.maybeNull(types.boolean) }), {}),
);

export const Values = types.model(
  Object.values(FIELDS).reduce((result, item) => ({ ...result, [item]: types.maybeNull(types.string) }), {}),
);
export const ErrorLabel = types.model(
  Object.values(FIELDS).reduce((result, item) => ({ ...result, [item]: types.maybeNull(types.string) }), {}),
);

export const Data = types.model({
  touched: Touched,
  values: Values,
  errorLabel: ErrorLabel,
});
export const Store = types
  .model({
    mounted: types.boolean,
    data: Data,
    disabled: types.boolean,
  })

  .views((self) => ({
    get values() {
      return getSnapshot(self.data.values);
    },
    get errorLabels() {
      return Object.values(FIELDS).reduce((result, item) => {
        if (!self.disabled && self.data.touched[item] && self.data.errorLabel[item]) {
          return {
            ...result,
            [item]: self.data.errorLabel[item],
          };
        }
        return result;
      }, {});
    },
    get errors() {
      return Object.values(FIELDS).reduce((result, item) => {
        if (!self.disabled && self.data.touched[item]) {
          return {
            ...result,
            [item]: self.data.values[item],
          };
        }
        return result;
      }, {});
    },
  }))

  .actions((self) => ({
    mount: () => {
      self.mounted = true;
    },

    unmount: () => {
      self.mounted = false;
      self.data = { values: {}, touched: {}, errorLabel: {} };
    },
    checkFieldsOnSave: () => {
      Object.values(FIELDS).forEach((item) => (self.data.touched[item] = true));
    },
    checkFillFields: (data) => {
      const checkFieldByNull = Object.values(data).filter((item) => item == null);
      const fillFields = checkFieldByNull.length > 0;
      return fillFields;
    },
    returnToDefault: () => {
      self.data.values[FIELDS.LOGIN] = null;
      self.data.values[FIELDS.PASSWORD] = null;
      self.data.values[FIELDS.EMAIL] = null;
      self.data.touched[FIELDS.LOGIN] = false;
      self.data.touched[FIELDS.PASSWORD] = false;
      self.data.touched[FIELDS.EMAIL] = false;
    },
    addNewUser: flow(function* addNewUser() {
      const root = getRoot(self);
      self.data.errorLabel = {};
      try {
        self.disabled = true;
        const data = {
          login: self.data.values[FIELDS.LOGIN],
          password: self.data.values[FIELDS.PASSWORD],
          email: self.data.values[FIELDS.EMAIL],
        };
        self.checkFieldsOnSave();
        const checkByNull = self.checkFillFields(data);
        if (checkByNull) {
          throw new SyntaxError('');
        }
        const result = yield root.api.authorizationStore.addNewUser(data);
        const { token } = result;
        const decoded = jwt(token);
        cookie.set('token', token, {
          expires: new Date(decoded.exp * 1000),
        });
        self.returnToDefault();
      } catch (error) {
        const errorMessage = error.response.data.message;
        if (error.name === 'SyntaxError') root.createNotificationMessage(LABELS.ERROR_FILL_FIELDS);
        if (errorMessage.includes(ERROR_LABELS.ERROR_EMAIL)) self.data.errorLabel[FIELDS.EMAIL] = LABELS.ERROR_EMAIL;
        if (errorMessage.includes(ERROR_LABELS.ERROR_LOGIN_MIN) || errorMessage.includes(ERROR_LABELS.ERROR_LOGIN_MAX))
          self.data.errorLabel[FIELDS.LOGIN] = LABELS.ERROR_LOGIN;
        if (errorMessage.includes(ERROR_LABELS.ERROR_PASSWORD))
          self.data.errorLabel[FIELDS.PASSWORD] = LABELS.ERROR_PASSWORD;
        if (errorMessage.includes(ERROR_LABELS.ERROR_PASSWORD_MAX))
          self.data.errorLabel[FIELDS.PASSWORD] = LABELS.ERROR_PASSWORD_MAX;
      } finally {
        self.disabled = false;
      }
    }),
    onChange: ({ id, value }) => {
      self.data.values[id] = value;
    },
    onFocus: () => {},

    onBlur: ({ id }) => {
      self.data.touched[id] = true;
    },
  }));

export function create() {
  return Store.create({
    mounted: false,
    data: { values: {}, touched: {}, errorLabel: {} },
    disabled: false,
  });
}
