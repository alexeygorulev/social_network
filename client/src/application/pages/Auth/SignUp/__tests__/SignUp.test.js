import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import SignUp from '../SignUp';

describe('Auth/SignUp', () => {
  const defaultProps = {
    mounted: true,
  };
  const addNewUserMock = jest.fn();
  const mountMock = jest.fn();
  const unmountMock = jest.fn();

  const setUp = (props = defaultProps) => {
    const { mounted } = props;
    return {
      mounted,
      mount: mountMock,
      unmount: unmountMock,
      values: '',
      onBlur: jest.fn(),
      onChange: jest.fn(),
      addNewUser: addNewUserMock,
      errorLabels: {},
      errors: {},
      disabled: false,
    };
  };
  it('snapshot with mounted=false', () => {
    const store = setUp();
    store.mounted = false;
    const wrapper = shallow(<SignUp store={store} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('snapshot with mounted=true', () => {
    const store = setUp();
    const onLogin = true;
    const wrapper = shallow(<SignUp store={store} onLogin={onLogin} />);
    expect(addNewUserMock).not.toHaveBeenCalled();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.find('button').simulate('click');
    expect(addNewUserMock).toHaveBeenCalled();
  });
});
