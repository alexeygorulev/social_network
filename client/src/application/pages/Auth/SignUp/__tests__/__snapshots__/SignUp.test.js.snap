// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth/SignUp snapshot with mounted=false 1`] = `""`;

exports[`Auth/SignUp snapshot with mounted=true 1`] = `
<div
  className="card__back"
>
  <div
    className="center__wrap"
  >
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <h4
        style={
          Object {
            "marginBottom": 5,
            "paddingBottom": 5,
          }
        }
      >
        Sign up
      </h4>
      <div
        className="form__group"
      >
        <Memo(InputText)
          id="PARAM_TYPES_ITEM_FIELD_LOGIN"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          placeholder="Your Login"
          type="text"
        />
        <p
          className="input__icon"
        >
          <span
            className="material-symbols-outlined"
          >
            person
          </span>
        </p>
      </div>
      <div
        className="form__group"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <Memo(InputText)
          id="PARAM_TYPES_ITEM_FIELD_EMAIL"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          placeholder="Your Email"
          type="email"
        />
        <p
          className="input__icon"
        >
          <span
            className="material-symbols-outlined"
          >
            alternate_email
          </span>
        </p>
      </div>
      <div
        className="form__group"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <Memo(InputText)
          id="PARAM_TYPES_ITEM_FIELD_PASSWORD"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          placeholder="Your Password"
          type="password"
        />
        <p
          className="input__icon"
        >
          <span
            className="material-symbols-outlined"
          >
            lock
          </span>
        </p>
      </div>
      <button
        className="btn"
        disabled={false}
        onClick={[MockFunction]}
        type="submit"
      >
        submit
      </button>
    </div>
  </div>
</div>
`;
